src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,365 @@
+        $resources = array();
+
+        if (false !== strpos($query, '*')) {
+            $resources = $this->iteratorToCollection($this->getGlobIterator($query));
+        } elseif ($this->store->exists($query)) {
+            $resources = array($this->store->get($query));
+        }
+
+        return new FilesystemResourceCollection($resources);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function contains($query, $language = 'glob')
+    {
+        if ('glob' !== $language) {
+            throw UnsupportedLanguageException::forLanguage($language);
+        }
+
+        Assert::stringNotEmpty($query, 'The glob must be a non-empty string. Got: %s');
+        Assert::startsWith($query, '/', 'The glob %s is not absolute.');
+
+        $query = Path::canonicalize($query);
+
+        if (false !== strpos($query, '*')) {
+            $iterator = $this->getGlobIterator($query);
+            $iterator->rewind();
+
+            return $iterator->valid();
+        }
+
+        return $this->store->exists($query);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function add($path, $resource)
+    {
+        Assert::stringNotEmpty($path, 'The path must be a non-empty string. Got: %s');
+        Assert::startsWith($path, '/', 'The path %s is not absolute.');
+
+        $path = Path::canonicalize($path);
+
+        if ($resource instanceof ResourceCollection) {
+            $this->ensureDirectoryExists($path);
+
+            foreach ($resource as $child) {
+                $this->addResource($path.'/'.$child->getName(), $child);
+            }
+
+            $this->sortStore();
+
+            return;
+        }
+
+        if ($resource instanceof FilesystemResource) {
+            $this->ensureDirectoryExists(Path::getDirectory($path));
+            $this->addResource($path, $resource);
+            $this->sortStore();
+
+            return;
+        }
+
+        throw new UnsupportedResourceException(sprintf(
+            'The passed resource must be a FilesystemResource or a ResourceCollection. Got: %s',
+            is_object($resource) ? get_class($resource) : gettype($resource)
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function remove($query, $language = 'glob')
+    {
+        $resources = $this->find($query, $language);
+        $nbOfResources = count($this->store->keys());
+
+        // Run the assertion after find(), so that we know that $query is valid
+        Assert::notEq('', trim($query, '/'), 'The root directory cannot be removed.');
+
+        foreach ($resources as $resource) {
+            $this->removeResource($resource);
+        }
+
+        return $nbOfResources - count($this->store->keys());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        $root = new GenericFilesystemResource('/');
+        $root->attachTo($this, '/');
+
+        // Subtract root
+        $removed = count($this->store->keys()) - 1;
```

Comment:
I'd change this to

```php
$removed = $this->store instanceof Countable
    ? count($this->store) - 1
    : count($this->store->keys()) - 1;
```

so that stores can provide optimized counting.